File: hw/villarreal/u3/hw21PythonAPI/product/product_controller.py
from fastapi import HTTPException
from sqlmodel import Session
from typing import Optional
from .models import ProductCreate, ProductUpdate, ProductUpdateStock
from .product_repository import product_repository

class ProductController:
    def get_all(self, session: Session, categoryId: Optional[int] = None):
        products = product_repository.find_all(session, categoryId)
        return {"success": True, "data": products}

    def get_by_id(self, session: Session, product_id: int):
        product = product_repository.find_by_id(session, product_id)
        if not product:
            raise HTTPException(status_code=404, detail="Product not found")
        return {"success": True, "data": product}

    def create(self, session: Session, product_data: ProductCreate):
        product = product_repository.create(session, product_data)
        return {"success": True, "data": product}

    def update(self, session: Session, product_id: int, product_data: ProductUpdate):
        product = product_repository.update(session, product_id, product_data)
        if not product:
            raise HTTPException(status_code=404, detail="Product not found")
        return {"success": True, "data": product}

    def update_stock(self, session: Session, product_id: int, stock_data: ProductUpdateStock):
        product = product_repository.update_stock(session, product_id, stock_data.stock)
        if not product:
            raise HTTPException(status_code=404, detail="Product not found")
        return {"success": True, "data": product}

    def remove(self, session: Session, product_id: int):
        success = product_repository.remove(session, product_id)
        if not success:
            raise HTTPException(status_code=404, detail="Product not found")
        return {"success": True, "message": "Product deleted"}

product_controller = ProductController()

File: exams/u3/andino/andino_ExamU3/BackEnd/Backend_Crud/src/controllers/ProductController.js
const ProductService = require('../services/ProductService');

const ProductController = {
  async getAll(req, res) {
    try {
      const products = await ProductService.getAll();
      res.json(products);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },

  async getById(req, res) {
    try {
      const product = await ProductService.getById(req.params.id);
      if (!product) return res.status(404).json({ error: 'Product not found' });
      res.json(product);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },

  async create(req, res) {
    try {
      const product = await ProductService.create(req.body);
      res.status(201).json(product);
    } catch (error) {
      res.status(400).json({ error: error.message });
    }
  },

  async update(req, res) {
    try {
      const product = await ProductService.update(req.params.id, req.body);
      res.json(product);
    } catch (error) {
      const status = error.message.includes('not found') ? 404 : 400;
      res.status(status).json({ error: error.message });
    }
  },

  async remove(req, res) {
    try {
      const product = await ProductService.remove(req.params.id);
      res.json({ message: 'Product deleted', product });
    } catch (error) {
      const status = error.message.includes('not found') ? 404 : 400;
      res.status(status).json({ error: error.message });
    }
  },
};

module.exports = ProductController;
